cunit/buf.testc

```diff
@@ -1654,6 +1655,218 @@ static void test_tocrlf(void)
+
+    /* each char in src is potentially four once escaped */
+    escaped = xzmalloc(4 * strlen(src) + 1);
+
+    for (dst = escaped; src && *src; dst++, src++) {
+        if (*src == '\033') {
+            strcpy(dst, "\\033");
+            dst += 3; /* plus one more from the loop iterator */
+        }
+        else {
+            *dst = *src;
+        }
+    }
+
+    return escaped;
+}
+
+#define MY_ASSERT_ANSI_STRING_EQUAL(actual,expected) do {               \
+    const char *_a = (actual), *_e = (expected);                        \
+    char *escaped_actual, *escaped_expected;                            \
+                                                                        \
+    escaped_actual = escape_ansi_string(_a);                            \
+    escaped_expected = escape_ansi_string(_e);                          \
+                                                                        \
+    CU_assertFormatImplementation(!strcmp(_a?_a:"",_e?_e:""), __LINE__, \
+        __FILE__, "", CU_FALSE,                                         \
+        "MY_ASSERT_ANSI_STRING_EQUAL(%s=\"%s\",%s=\"%s\")",             \
+        #actual, escaped_actual, #expected, escaped_expected);          \
+                                                                        \
+    free(escaped_actual);                                               \
+    free(escaped_expected);                                             \
+} while (0);
+
+static void test_append_sgr(void)
+{
+    struct {
+        int in_values[5];
+        char *expect;
+    } testcases[] = {
+        { { SGR_DONE },             "\033[m" },
+        { { -123, SGR_DONE },       "\033[m" },
+        { { 0, SGR_DONE },          "\033[0m" },
+        { { 33, SGR_DONE },         "\033[33m" },
+        { { 38, 5, 127, SGR_DONE }, "\033[38;5;127m" },
+        { { 0, SGR_DONE, 2, 3, 4 }, "\033[0m" },
+    };
+    const size_t n_testcases = sizeof(testcases) / sizeof(testcases[0]);
+    size_t i;
+    struct buf buf = BUF_INITIALIZER;
+
+    for (i = 0; i < n_testcases; i++) {
+        buf_reset(&buf);
+
+        buf_append_sgr(&buf, testcases[i].in_values[0],
+                             testcases[i].in_values[1],
+                             testcases[i].in_values[2],
+                             testcases[i].in_values[3],
+                             testcases[i].in_values[4],
+                             SGR_DONE /* belt and suspenders */);
+
+        MY_ASSERT_ANSI_STRING_EQUAL(buf_cstring(&buf), testcases[i].expect);
+    }
+
+    buf_free(&buf);
+}
+
+static void test_append_kv_basic(void)
+{
+    struct {
+        int   in_sep;
+        char *in_key;
+        char *in_value;
+        char *expect;
+    } testcases[] = {
+        { 0,   NULL,    NULL,       "" },
+        { ' ', NULL,    NULL,       " " },
+        { 0,   "mykey", NULL,       "mykey=<>" },
+        { ' ', "mykey", NULL,       " mykey=<>" },
+        { 0,   NULL,    "myvalue",  "myvalue" },
```

Comment:
It's not visible in this test because it doesn't use the `want_color` argument, but the `key=NULL, value="myvalue" => "myvalue"` case is useful for when you want to output a deterministically coloured string that isn't really a key-value pair (and thus makes no sense to have `=<>`) -- i.e. just a value.  IIRC I think I used something like this for the "CAL" etc type abbreviations in the pretty output.

If we extend the deterministically-coloured output behaviour to non-kv functions, this will become redundant, but for the moment it saves me inventing another API...  and it's the behaviour that made the most sense to me when I was iterating all the possible argument combinations for the test case.  If we had a separate function for a single deterministically coloured string, and made this function strictly for kv pairs, then I think a NULL key should probably be regarded as a programmer error and throw an assertion...

The `struct buf` API generally doesn't differentiate between NULL and empty string, so I won't make things weird by introducing that behaviour in just these functions, it would just make them easier to hold wrong.